sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,363 @@
+Mutating admission webhooks are needed to support parity in the chain, so we know we'll eventually
+need them and because they will directly interact in the registration API, we need to know how we'll
+build them before we can push a registration API to beta.
+
+
+### Admission interfaces
+
+The current interface in `pkg/admission/interfaces.go` looks like this:
+
+```go
+type Interface interface {
+  Admit(a Attributes) (err error)
+  Handles(operation Operation) bool
+}
+```
+
+We'd like to break this into multiple interfaces, one per phase, called at the appropriate times
+within the request lifecycle.
+
+```go
+type Mutator interface {
+  // ApplyMutations is executed after the request body is deserialized into an
+  // object for create and update requests.
+  // Plugins are free to modify the object as they see fit. If an error is
+  // returned, the request is marked as failed and no changes are persisted.
+  ApplyMutations(a Attributes) error
+}
+
+type Validator interface {
+  // Validate is executed after the appropriate `rest.Storage` handler's
+  // validation occurs. This gives the plugin the opportunity to validate the
+  // object after the handler has validated it, but before it is persisted. If
+  // an error is returned, the request is marked as failed and no changes are
+  // persisted.
+  Validate(a Attributes) error
+}
+```
+
+### rest Storage
+
+To be able to invoke admission plugin validation at the right time, we'll need
+to modify the various `rest.Storage` interfaces, as described below.
+
+#### Creater, NamedCreater
+
+- add `BeforeCreate(ctx api.Context, obj runtime.Object) error`
+- add `Validate(ctx api.Context, obj runtime.Object) error`
+
+#### Updater, CreaterUpdater
+
+- add `BeforeUpdate(ctx api.Context, obj, old runtime.Object) error`
+- add `ValidateUpdate(ctx api.Context, obj, old runtime.Object) field.ErrorList`
+
+#### Patcher
+
+- inherits changes from `Updater`
+
+
+#### Deleter, GracefulDeleter
+
+- add `BeforeDelete(ct api.Context, obj runtime.Object, options *api.DeleteOptions) (graceful, gracefulPending bool, err error)`
+
+#### Connector
+
+- QUESTION: do we need to adjust this interface?
+
+
+## REST flows
+
+### Create
+
+#### Current behavior
+
+This is the current [apiserver create
+flow](https://github.com/kubernetes/kubernetes/blob/ef0c9f0c5b8efbba948a0be2c98d9d2e32e0b68c/pkg/apiserver/resthandler.go#L333):
+
+1. decode incoming request
+1. admit.Admit
+1. r.Create (the items below are part of the generic registry store / strategy-based storage, `pkg/registry/generic/registry/store.go`)
+  1. rest.BeforeCreate
+    1. strategy.PrepareForCreate
+    1. api.FillObjectMetaSystemFields
+    1. api.GenerateName
+    1. strategy.Validate
+    1. validation.ValidateObjectMeta
+    1. strategy.Canonicalize
+  1. e.Storage.Create
+  1. e.AfterCreate
+  1. e.Decorator
+
+#### Proposed behavior
+
+This is the proposed new behavior:
+
+1. decode request
+1. admit.ApplyMutations
```

Comment:
At this point the object will be of the internal version. We should send the external version objects to the webhooks. I think it should be required for beta, otherwise webhooks will be hard to maintain.